`azurerm_container_app` -  Add `key_vault_url` and `identity_id` for `secret` nested block.

lonegunmanb opened the pull request:
We've received a feature request [here](https://github.com/Azure/terraform-azure-container-apps/issues/52), container app now supports [reference key vault secret via url](https://learn.microsoft.com/en-us/azure/container-apps/manage-secrets?tabs=arm-template#reference-secret-from-key-vault) now.

This pull request added support for `key_vault_url` and `identity_id` for `secret` nested block. I found that once a `secret` block was added with `key_vault_url`, the value of secret would be returned while read, if we set this secret value back to the state then we'll have a configuration drift. So this pr set secret's value to empty string when `key_vault_url` is not empty in the response.

Because only absolute attribute paths, ones starting with top level attribute names, are supported by the `ConflictsWith`, attribute paths cannot be accurately declared for TypeList (if MaxItems is greater than 1), TypeMap, or TypeSet attributes, this pr didn't declare `ConflictsWith`, but check whether users have set both of `key_vault_url` and `value` in `Creat` and `Update` methods.

tombuildsstuff commented on internal/services/containerapps/helpers/container_apps.go:
```diff
@@ -2566,10 +2569,24 @@ func SecretsSchema() *pluginsdk.Schema {
+				"key_vault_url": {
+					Type:         pluginsdk.TypeString,
+					Optional:     true,
+					ValidateFunc: validation.IsURLWithHTTPS,
```
can we validate that this is a Key Vault Nested Item, since we don't support Managed HSM Nested Items in the provider at this time?

lonegunmanb commented:
Hi @tombuildsstuff thanks for the review, I've added a new validator to ensure that it's an url for key vault key, all validate dns suffixes could be found [here](https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#dns-suffixes-for-object-identifiers), and I've also added an unit test to check this rule. Would you please give this pr another review? Thanks!

lonegunmanb commented:
@GurliGebis thanks for the tip, I've also found that #23958's author @x-delfino will not be able to continue that pr for a while, I'd like to continue his work via this pr. I'll try to rebase this pr to his branch and try again.

GurliGebis commented:
@lonegunmanb have you looked at the reviews done to the other PR? (I think there where some changes in there, that is needed - so would make good sense to include them here)

lonegunmanb commented:
> @lonegunmanb have you looked at the reviews done to the other PR? (I think there where some changes in there, that is needed - so would make good sense to include them here)

Yes, I've read comments left by @magodo and @jackofallops . Now this pr followed what @x-delfino left behind (I think his approach was very close to a success)

falsadoo commented:
@lonegunmanb can review the comments for this? We really need feature for our project

lonegunmanb commented on internal/services/containerapps/helpers/container_apps.go:
```diff
@@ -2599,21 +2631,36 @@ func SecretsDataSourceSchema() *pluginsdk.Schema {
+		return fmt.Errorf("must supply key vault secret id when specifying identity")
+	}
+	return nil
+}
```
We can't use `RequiredWith` because `secret` is a `Set`. I've moved the validation logic into `CustomizeDiff`.

lonegunmanb commented:
@stephybun I've updated the pr as you suggested and ran the tests, would you please give this pr another review? Thanks!:

=== RUN   TestAccContainerAppResource_basic
=== PAUSE TestAccContainerAppResource_basic
=== CONT  TestAccContainerAppResource_basic
--- PASS: TestAccContainerAppResource_basic (591.41s)
=== RUN   TestAccContainerAppResource_workloadProfile
=== PAUSE TestAccContainerAppResource_workloadProfile
=== CONT  TestAccContainerAppResource_workloadProfile
--- PASS: TestAccContainerAppResource_workloadProfile (1844.85s)
=== RUN   TestAccContainerAppResource_smallerGranularityCPUMemoryCombinations
=== PAUSE TestAccContainerAppResource_smallerGranularityCPUMemoryCombinations
=== CONT  TestAccContainerAppResource_smallerGranularityCPUMemoryCombinations
--- PASS: TestAccContainerAppResource_smallerGranularityCPUMemoryCombinations (2079.20s)
=== RUN   TestAccContainerAppResource_workloadProfileUpdate
=== PAUSE TestAccContainerAppResource_workloadProfileUpdate
=== CONT  TestAccContainerAppResource_workloadProfileUpdate
--- PASS: TestAccContainerAppResource_workloadProfileUpdate (3340.86s)
=== RUN   TestAccContainerAppResource_withSystemAssignedIdentity
=== PAUSE TestAccContainerAppResource_withSystemAssignedIdentity
=== CONT  TestAccContainerAppResource_withSystemAssignedIdentity
--- PASS: TestAccContainerAppResource_withSystemAssignedIdentity (743.44s)
=== RUN   TestAccContainerAppResource_withUserAssignedIdentity
=== PAUSE TestAccContainerAppResource_withUserAssignedIdentity
=== CONT  TestAccContainerAppResource_withUserAssignedIdentity
--- PASS: TestAccContainerAppResource_withUserAssignedIdentity (646.38s)
=== RUN   TestAccContainerAppResource_withSystemAndUserAssignedIdentity
=== PAUSE TestAccContainerAppResource_withSystemAndUserAssignedIdentity
=== CONT  TestAccContainerAppResource_withSystemAndUserAssignedIdentity
--- PASS: TestAccContainerAppResource_withSystemAndUserAssignedIdentity (737.31s)
=== RUN   TestAccContainerAppResource_withIdentityUpdate
=== PAUSE TestAccContainerAppResource_withIdentityUpdate
=== CONT  TestAccContainerAppResource_withIdentityUpdate
--- PASS: TestAccContainerAppResource_withIdentityUpdate (1492.30s)
=== RUN   TestAccContainerAppResource_withKeyVaultSecretVersioningUpdate
=== PAUSE TestAccContainerAppResource_withKeyVaultSecretVersioningUpdate
=== CONT  TestAccContainerAppResource_withKeyVaultSecretVersioningUpdate
--- PASS: TestAccContainerAppResource_withKeyVaultSecretVersioningUpdate (882.54s)
=== RUN   TestAccContainerAppResource_withKeyVaultSecretIdentityUpdate
=== PAUSE TestAccContainerAppResource_withKeyVaultSecretIdentityUpdate
=== CONT  TestAccContainerAppResource_withKeyVaultSecretIdentityUpdate
--- PASS: TestAccContainerAppResource_withKeyVaultSecretIdentityUpdate (1339.02s)
=== RUN   TestAccContainerAppResource_basicUpdate
=== PAUSE TestAccContainerAppResource_basicUpdate
=== CONT  TestAccContainerAppResource_basicUpdate
--- PASS: TestAccContainerAppResource_basicUpdate (781.08s)
=== RUN   TestAccContainerAppResource_requiresImport
=== PAUSE TestAccContainerAppResource_requiresImport
=== CONT  TestAccContainerAppResource_requiresImport
--- PASS: TestAccContainerAppResource_requiresImport (598.95s)
=== RUN   TestAccContainerAppResource_complete
=== PAUSE TestAccContainerAppResource_complete
=== CONT  TestAccContainerAppResource_complete
--- PASS: TestAccContainerAppResource_complete (693.93s)
=== RUN   TestAccContainerAppResource_completeVolumeEmptyDir
=== PAUSE TestAccContainerAppResource_completeVolumeEmptyDir
=== CONT  TestAccContainerAppResource_completeVolumeEmptyDir
--- PASS: TestAccContainerAppResource_completeVolumeEmptyDir (754.19s)
=== RUN   TestAccContainerAppResource_completeWithNoDaprAppPort
=== PAUSE TestAccContainerAppResource_completeWithNoDaprAppPort
=== CONT  TestAccContainerAppResource_completeWithNoDaprAppPort
--- PASS: TestAccContainerAppResource_completeWithNoDaprAppPort (717.00s)
=== RUN   TestAccContainerAppResource_completeWithVNet
=== PAUSE TestAccContainerAppResource_completeWithVNet
=== CONT  TestAccContainerAppResource_completeWithVNet
--- PASS: TestAccContainerAppResource_completeWithVNet (1222.30s)
=== RUN   TestAccContainerAppResource_completeWithSidecar
=== PAUSE TestAccContainerAppResource_completeWithSidecar
=== CONT  TestAccContainerAppResource_completeWithSidecar
--- PASS: TestAccContainerAppResource_completeWithSidecar (739.71s)
=== RUN   TestAccContainerAppResource_completeWithMultipleContainers
=== PAUSE TestAccContainerAppResource_completeWithMultipleContainers
=== CONT  TestAccContainerAppResource_completeWithMultipleContainers
--- PASS: TestAccContainerAppResource_completeWithMultipleContainers (713.56s)
=== RUN   TestAccContainerAppResource_completeUpdate
=== PAUSE TestAccContainerAppResource_completeUpdate
=== CONT  TestAccContainerAppResource_completeUpdate
--- PASS: TestAccContainerAppResource_completeUpdate (815.53s)
=== RUN   TestAccContainerAppResource_completeTcpExposedPort
=== PAUSE TestAccContainerAppResource_completeTcpExposedPort
=== CONT  TestAccContainerAppResource_completeTcpExposedPort
--- PASS: TestAccContainerAppResource_completeTcpExposedPort (1203.16s)
=== RUN   TestAccContainerAppResource_removeDaprAppPort
=== PAUSE TestAccContainerAppResource_removeDaprAppPort
=== CONT  TestAccContainerAppResource_removeDaprAppPort
--- PASS: TestAccContainerAppResource_removeDaprAppPort (807.73s)
=== RUN   TestAccContainerAppResource_secretFail
=== PAUSE TestAccContainerAppResource_secretFail
=== CONT  TestAccContainerAppResource_secretFail
--- PASS: TestAccContainerAppResource_secretFail (727.38s)
=== RUN   TestAccContainerAppResource_scaleRules
=== PAUSE TestAccContainerAppResource_scaleRules
=== CONT  TestAccContainerAppResource_scaleRules
--- PASS: TestAccContainerAppResource_scaleRules (653.58s)
=== RUN   TestAccContainerAppResource_multipleScaleRules
=== PAUSE TestAccContainerAppResource_multipleScaleRules
=== CONT  TestAccContainerAppResource_multipleScaleRules
--- PASS: TestAccContainerAppResource_multipleScaleRules (690.53s)
=== RUN   TestAccContainerAppResource_scaleRulesUpdate
=== PAUSE TestAccContainerAppResource_scaleRulesUpdate
=== CONT  TestAccContainerAppResource_scaleRulesUpdate
--- PASS: TestAccContainerAppResource_scaleRulesUpdate (871.02s)
=== RUN   TestAccContainerAppResource_ipSecurityRulesUpdate
=== PAUSE TestAccContainerAppResource_ipSecurityRulesUpdate
=== CONT  TestAccContainerAppResource_ipSecurityRulesUpdate
--- PASS: TestAccContainerAppResource_ipSecurityRulesUpdate (915.85s)
=== RUN   TestAccContainerAppResource_ingressTrafficValidation
=== PAUSE TestAccContainerAppResource_ingressTrafficValidation
=== CONT  TestAccContainerAppResource_ingressTrafficValidation
--- PASS: TestAccContainerAppResource_ingressTrafficValidation (14.25s)
PASS

lonegunmanb commented:
> @lonegunmanb can you review the comments for this? We really need feature for our project

@falsadoo sorry for the late update, I think the latest try would pass the review. Thanks for asking.

stephybun reviewed (approved):
Thanks @lonegunmanb LGTM 🦆
